### PR TITLE
2328 course study mode question publish trainees

### PR DESCRIPTION
**Pull request by avinhurry**
### Context

Adding the course study mode question to Publish courses. When a user chooses a Publish course which has the `study_mode`: `full_time_or_part_time`, we add an extra step in the user journey asking the user whether they want the `full_time` or `part_time` course.

### Changes proposed in this pull request

- Added `course_study_mode` workflow
- Added UI page for choosing `study_mode`
- Added migration for `study_mode` column on `Courses` table
- Added logic in order for this to only be applied to specific routes
- Added validation to ensure the user chooses `full_time` or `part_time`

### Guidance to review

- Navigate to a trainee on one of the following routes: `provider_led_postgrad`, `school_direct_tuition_fee` or `school_direct_salaried`
- Go to the course details page where you can select a Publish course
- Select a Publish course which is `Full time or Part time` and click `Continue`
- Choose `Full time` or `Part time` and click `Continue`
- Check that the selected `study_mode` appears on the summary page 
- Click `Confirm` and check that the progress is `Completed`

### CODE REVIEW

**Comment by avinhurry** [comment]
[[L] Course study mode question - Publish trainees](https://trello.com/c/XBDN1FnL/2328-l-course-study-mode-question-publish-trainees)

**Review by mooktakim — CHANGES_REQUESTED** [review]
Couple of comments

File: app/controllers/concerns/publish_course_next_path.rb [hunk 1]
```diff
@@ -17,4 +18,20 @@ def course_confirmation_path
...
+    @course ||= trainee.available_courses.find_by_code!(course_code)
+  end
+
+  def publish_course_details_form
+    @publish_course_details_form ||= PublishCourseDetailsForm.new(trainee)
+  end
+
+  def requires_study_mode?
+    course.present? && trainee.requires_study_mode? && course.study_mode == "full_time_or_part_time"
+  end
```
[the remark this answers is not in the document]
Comment (avinhurry) [reply app/controllers/concerns/publish_course_next_path.rb:36]: study_mode will never be nil

**Review by mooktakim — APPROVED** [review]
Looks good

**Comment by bbelward** [comment]
Hey @avinhurry, I'm getting the form step, but study mode doesn't show up on the confirm page 😢  Tried it on multiple courses, also varying between full time and part time options. 
2328studymode

**Comment by avinhurry** [comment]
🤔 @bbelward, I can't seem to replicate that. Can you let me know when you have time for a chat please?

**Pull request merged**